pkg/resource/firewall/hook.go

```diff
@@ -27,3 +36,251 @@ var (
+	keys := make([]string, 0, len(cfg.LogDestination))
+	for k := range cfg.LogDestination {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	key := "LogDestination:"
+	for _, k := range keys {
+		v := *cfg.LogDestination[k]
+		key += fmt.Sprintf("%s:%s,", k, v)
+	}
+
+	if cfg.LogDestinationType != nil {
+		key += fmt.Sprintf("LogDestinationType:%s,", *cfg.LogDestinationType)
+	}
+	if cfg.LogType != nil {
+		key += fmt.Sprintf("LogType:%s", *cfg.LogType)
+	}
+	return key
+}
+
+func makeLogDestinationConfigMap(logDestinationConfig []*svcapitypes.LogDestinationConfig) map[string]*svcapitypes.LogDestinationConfig {
+	logDestinationConfigMap := make(map[string]*svcapitypes.LogDestinationConfig)
+
+	for _, cfg := range logDestinationConfig {
+		logDestinationConfigMap[makeLogDestinationConfigMapKey(cfg)] = cfg
+	}
+	return logDestinationConfigMap
+}
+
+// compareLoggingDestinationConfigs generates LogDestinationConfigs which need to be
+// added and deleted to reach desired LogDestinationConfig from latest LogDestinationConfig
+func compareLoggingDestinationConfigs(desired []*svcapitypes.LogDestinationConfig, latest []*svcapitypes.LogDestinationConfig) (add []*svcapitypes.LogDestinationConfig, delete []*svcapitypes.LogDestinationConfig) {
+	add = make([]*svcapitypes.LogDestinationConfig, 0)
+	delete = make([]*svcapitypes.LogDestinationConfig, 0)
+
+	desiredMap := makeLogDestinationConfigMap(desired)
+	latestMap := makeLogDestinationConfigMap(latest)
+
+	for key, val := range desiredMap {
+		if _, ok := latestMap[key]; !ok {
+			add = append(add, val)
+		}
+	}
+
+	for key, val := range latestMap {
+		if _, ok := desiredMap[key]; !ok {
+			delete = append(delete, val)
+		}
+	}
+
+	return add, delete
+}
+
+// syncLoggingConfiguration gets desired logging config and latest (existing)
+// logging config as input. It compares both and applies the delta to ensure
+// desired logging config is configured for the firewall.
+func (rm *resourceManager) syncLoggingConfiguration(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncLoggingConfiguration")
+	defer exit(err)
+
+	var input *svcsdk.UpdateLoggingConfigurationInput
+	var add, delete []*svcapitypes.LogDestinationConfig
+	if latest != nil && desired != nil {
+		add, delete = compareLoggingDestinationConfigs(desired.ko.Spec.LoggingConfiguration.LogDestinationConfigs, latest.ko.Spec.LoggingConfiguration.LogDestinationConfigs)
+		input = rm.newLoggingConfigurationPayload(latest)
+	} else if latest == nil {
+		add = desired.ko.Spec.LoggingConfiguration.LogDestinationConfigs
+		input = rm.newLoggingConfigurationPayload(desired)
+		input.LoggingConfiguration = &svcsdk.LoggingConfiguration{}
+	} else if desired == nil {
+		delete = latest.ko.Spec.LoggingConfiguration.LogDestinationConfigs
+		input = rm.newLoggingConfigurationPayload(latest)
```

Comment:
Will desired ever be nil?
I think this part can be simplified assuming that if we ever reach this area desired != nil

Reply:
yes. When loggingConfiguration is removed, desired is nil.

Reply:
done